nrf: Add support for nrf52820

Mierunski opened the pull request:
Add support for nrf52820 soc and emulated development for nrf52820 on nrf52833dk

Mierunski committed:
soc: nordic: Add support for nRF52820

This commit adds basic support for nRF52820 SoC.
Changes affect introducing:
 - architecuture files (dtsi)
 - configuration of nrfx drivers
 - adaptation of inclusions based on chosen SoC

Signed-off-by: Mieszko Mierunski <mieszko.mierunski@nordicsemi.no>

Mierunski committed:
boards: nordic: Add support for nRF52833dk_nrf52820 board

This commit adds support for nRF52820 development on nRF52833DK.
Changes afffects:
 - Introduce files related to board description.
 - Add blank documentation file (for future update).
 - configuration files for build process.

Signed-off-by: Mieszko Mierunski <mieszko.mierunski@nordicsemi.no>

anangl commented:
@Mierunski You need to add a change in `west.yml` that will point to `pull/37/head` as the hal_nordic module revision to be used.

sigurdnev commented on dts/arm/nordic/nrf52820.dtsi:
```diff
@@ -0,0 +1,263 @@
+			status = "disabled";
+			label = "QDEC";
+		};
+
```
rng: random@4000d000 {
			compatible = "nordic,nrf-rng";
			reg = <0x4000d000 0x1000>;
			interrupts = <13 1>;
			status = "okay";
			label = "RNG";
		};

lemrey commented on boards/arm/nrf52833dk_nrf52820/Kconfig.board:
```diff
@@ -0,0 +1,8 @@
+# nRF52833DK nRF52820 board configuration
+
+# Copyright (c) 2019 Nordic Semiconductor ASA
```
2020

lemrey commented on boards/arm/nrf52833dk_nrf52820/nrf52833dk_nrf52820.dts:
```diff
@@ -0,0 +1,164 @@
+
+/ {
+	model = "Nordic nRF52833 DK NRF52820";
+	compatible = "nordic,nrf52833-dk-nrf52820", "nordic,nrf52820-qdaa",
```
There are some recent changes related to `compatible` here: https://github.com/zephyrproject-rtos/zephyr/pull/24774.

anangl commented on boards/arm/nrf52833dk_nrf52820/nrf52833dk_nrf52820.yaml:
```diff
@@ -0,0 +1,20 @@
+ram: 32
+flash: 256
+supported:
+  - arduino_gpio
```
Since `arduino_header` is not defined in dts for this board, this entry should not be present here.

anangl commented on dts/arm/nordic/nrf52820.dtsi:
```diff
@@ -0,0 +1,271 @@
+#include <dt-bindings/gpio/gpio.h>
+#include "nrf5_common.dtsi"
+
+/ {
```
Add the following here:
```
	chosen {
		zephyr,entropy = &rng;
		zephyr,flash-controller = &flash_controller;
	};
```

anangl commented on dts/arm/nordic/nrf52820.dtsi:
```diff
@@ -0,0 +1,271 @@
+
+	soc {
+
+		flash-controller@4001e000 {
```
Add `flash_controller: `.

anangl commented on dts/arm/nordic/nrf52820.dtsi:
```diff
@@ -0,0 +1,271 @@
+			label = "USBD";
+		};
+
+		wdt: watchdog@40010000 {
```
Add `wdt0: `.

nvlsianpu commented on boards/arm/nrf52833dk_nrf52820/nrf52833dk_nrf52820.dts:
```diff
@@ -0,0 +1,164 @@
+			label = "image-1";
+			reg = <0x0003E000 0x32000>;
+		};
+		scratch_partition: partition@70000 {
```
remove the scratch - mcuboot has new algorithm swap which can live without it.

carlescufi commented on west.yml:
```diff
@@ -53,7 +53,7 @@ manifest:
       path: modules/hal/cypress
     - name: hal_nordic
-      revision: 1ebfdf148b2ca1d9ab9267ab2cd64cc00f32d505
+      revision: pull/37/head
```
Should this be changed in the SoC commit instead?

Mierunski replied:
Yes, I will change it to proper hash when https://github.com/zephyrproject-rtos/hal_nordic/pull/37 is merged

ioannisg replied:
I guess the comment here is to do the change in the commit that contributes the SoC definition :)

carlescufi commented:
@Mierunski please rebase, the failure in CI was due to an unrelated issue

sigurdnev commented on dts/arm/nordic/nrf52820.dtsi:
```diff
@@ -0,0 +1,277 @@
+
+		cpu@0 {
+			device_type = "cpu";
+			compatible = "arm,cortex-m4f";
```
FPU is not present in nRF52820.
change to compatible = "arm,cortex-m4";

sigurdnev commented on soc/arm/nordic_nrf/nrf52/Kconfig.soc:
```diff
@@ -92,6 +92,59 @@ config SOC_NRF52811
+config SOC_NRF52820
+	depends on SOC_SERIES_NRF52X
+	bool
+	select CPU_CORTEX_M_HAS_DWT
```
Data watchpoint and trace is not present in nRF52820

sigurdnev commented on soc/arm/nordic_nrf/nrf52/Kconfig.soc:
```diff
@@ -92,6 +92,59 @@ config SOC_NRF52811
+	depends on SOC_SERIES_NRF52X
+	bool
+	select CPU_CORTEX_M_HAS_DWT
+	select CPU_HAS_FPU
```
FPU is not present in nRF52820

Mierunski committed:
bluetooth: Add bluetooth support for nRF52820

This commit adds support for bluetooth in nRF52820 SoC.
Bluetooth radio related files created and added to Zephyr.

Signed-off-by: Mieszko Mierunski <mieszko.mierunski@nordicsemi.no>

ioannisg reviewed (changes requested):
did a pass; no more comments except the one mentioned by @anangl  and  @sigurdnev

carlescufi commented:
@Mierunski and @anangl if the nrfx update is backwards compatible, let's have it as its own, separate, first commit.

Mierunski committed:
hal: nrf: Integrate nRFX 2.2.0 release

Change west.yml nordic hal module to point to 2.2.0 nrfx version.

Signed-off-by: Mieszko Mierunski <mieszko.mierunski@nordicsemi.no>

ioannisg commented on soc/arm/nordic_nrf/nrf52/Kconfig.defconfig.nrf52820_QDAA:
```diff
@@ -0,0 +1,19 @@
+	int
+	default 40
+
+config NET_CONFIG_IEEE802154_DEV_NAME
```
Is this needed now? HAS_HW_NRF_RADIO_802154 was removed

sigurdnev replied:
Why was HAS_HW_NRF_RADIO_IEEE802154 removed ?
nRF52820 radio supports IEEE 802.15.4

ioannisg replied:
i believe that's because the driver is not supporting the SoC yet, but maybe i am wrong

rlubos replied:
@ioannisg is correct, the driver does not support this SoC yet

sigurdnev replied:
Ok, that makes sense. Thanks for the explanation.

ioannisg reviewed (approved):
Looks good to me, now, thanks @Mierunski

sigurdnev commented:
Why was HAS_HW_NRF_RADIO_IEEE802154 removed ?
nRF52820 radio supports IEEE 802.15.4

Mierunski commented:
@ioannisg All green
